src/wifi/model/periodic-dmg-wifi-scheduler.cc

```diff
@@ -335,79 +293,97 @@ PeriodicDmgWifiScheduler::GetAvailableBlocks (uint32_t allocDuration, uint32_t s
 }
 
 void
-PeriodicDmgWifiScheduler::UpdateAvailableSlots (uint32_t startAlloc, uint32_t endAlloc, uint32_t difference)
+PeriodicDmgWifiScheduler::UpdateAvailableSlots (uint32_t startAlloc, uint32_t endAlloc)
 {
-  NS_LOG_FUNCTION (this);
+  NS_LOG_FUNCTION (this << startAlloc << endAlloc);
 
-  std::vector<std::pair<uint32_t, uint32_t> > newDti;
+  uint32_t startSlot;
+  uint32_t endSlot;
 
-  if (difference > 0)
+  for (auto it = m_availableSlots.begin(); it != m_availableSlots.end(); ++it)
     {
-      // something has changed in the allocation list, need to change the list of 
-      // available slots accordingly
-      bool search = true;
-      for (const auto & slot: m_availableSlots)
+      startSlot = (*it).first;
+      endSlot = (*it).second;
+      
+      if (startSlot > endAlloc || endSlot < startAlloc)
         {
+          continue;
+        }
 
-          if (slot.first < startAlloc)
-            {
-              newDti.push_back (slot);
-              continue;
-            }
-
-          if (endAlloc < slot.first)
-            {
-              // the following snippet ensure that the search is carried on until 
-              // we find the empty gap created by the allocation time reduction
-              // and we add it to the list of available slots 
-              if (search)
-                {
-                  if (difference == (slot.first - endAlloc))
-                    {
-                      // this make sure that two adjacent available slots are merged 
-                      newDti.push_back (std::make_pair (endAlloc, slot.second));
-                      search = false;
-                    }
-                  else if (difference < (slot.first - endAlloc))
-                    {
-                      // this condition is verified when there is one or more allocations 
-                      // between the gap and the following available slot
-                      newDti.push_back (std::make_pair (endAlloc, endAlloc + difference));
-                      newDti.push_back (slot);
-                      search = false;
-                    }
-                  NS_ASSERT_MSG (difference <= (slot.first - endAlloc), "Something broke in runtime, check the update of the available slots.");
-                }
-              else
-                {
-                  newDti.push_back (slot);
-                }
-            }
+      if (startSlot == startAlloc)
+        {
+          m_availableSlots.insert(it, std::make_pair (endAlloc, endSlot));
+          it = m_availableSlots.erase(it);
```

Comment:
Doesn't this work? It would be easier to read and more efficient as well (need reference iterator)
```suggestion
          it->first = endAlloc;
          ++it;
```